pkg/roachprod/roachprod.go

```diff
@@ -1330,3 +1331,58 @@ func InitProviders() map[string]string {
 
 	return providersState
 }
+
+// InitGrafana spins up a promethius and grafana instance on the provided nodes
+func InitGrafana(
+	ctx context.Context, l *logger.Logger, clusterName string, grafanaURL string,
+) error {
+	if err := LoadClusters(); err != nil {
+		return err
+	}
+	c, err := newCluster(l, clusterName)
+	if err != nil {
+		return err
+	}
+	nodes := c.TargetNodes()
+
+	ips, err := IP(ctx, l, clusterName, true)
+	if err != nil {
+		return err
+	}
+
+	var promCfg prometheus.Config
+	promCfg.WithPrometheusNode(nodes[0])
+	if err := promCfg.WithCluster(nodes, ips); err != nil {
+		return err
+	}
+	if err := promCfg.WithNodeExporter(nodes, ips); err != nil {
+		return err
+	}
+	promCfg.WithGrafanaDashboard(grafanaURL)
+
+	promNodeClusterName := strings.Split(clusterName, ":")[0] + ":" + fmt.Sprint(nodes[0])
+	promC, err := newCluster(l, promNodeClusterName)
+	if err != nil {
+		return err
+	}
+	_, err = prometheus.Init(ctx, l, c, promC, promCfg)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func StopGrafana(ctx context.Context, l *logger.Logger, clusterName string) error {
+	if err := LoadClusters(); err != nil {
+		return err
+	}
+	c, err := newCluster(l, clusterName)
+	if err != nil {
+		return err
+	}
+	nodes := c.TargetNodes()
```

Comment:
Instead of using c.TargetNodes(), we should use install.ListNodes. If the user passes in `roachprod stop-grafana $CLUSTER:2-3` where $CLUSTER is a 5 node cluster, we'll try to muck with node 2 or 3, when IMO we should only allow running this monitoring server on $CLUSTER:5. Making this fool proof entails taking just the cluster name and using it to figure out all nodes in it in order to target a fixed one. This comment applies equally to StartGrafana, and is patched in the top-level diff.

Reply:
> Instead of using c.TargetNodes(), we should use install.ListNodes. If the user passes in roachprod stop-grafana $CLUSTER:2-3 where $CLUSTER is a 5 node cluster, we'll try to muck with node 2 or 3, when IMO we should only allow running this monitoring server on $CLUSTER:5. Making this fool proof entails taking just the cluster name and using it to figure out all nodes in it in order to target a fixed one. This comment applies equally to StartGrafana, and is patched in the top-level diff.

We should be careful here. A convention in roachprod is to put the workload on the last node (and so to put the monitoring there) but I don't think this is good enough to hard-code restrictions into roachprod. I would be in favor of preventing "shotgun installation" of the monitoring stack though, since that's likely user error.

Of course you do want to install node_exporter on most nodes, so the reasoning doesn't hold for that step.

Reply:
I'd be ok with a flag to let users manually specify which node to run the monitoring stack from, but I rarely see myself or others needing to use it, so don't care about it being introduced in this PR. The UX I'm ultimately after is just `roachprod {start,stop}-grafana <cluster name>` and being able to access it through `roachprod grafanaurl <cluster name> --open` without needing to care about where the monitoring stack is running.

Reply:
Currently, only the `targetNodes()` get scraped, even for node_exporter. Is that okay for this PR? In a future PR, we could add the flags `--node-exporter $CLUSTER1-3` and `--crl-scrape $CLUSTER3-4` if we really wanted it.

Reply:
Revision: the roachprod grafana calls assume that all available nodes will be scraped and will set up the prom/grafana servers on the highest numbered node. This implies that `roachprod {start,stop}-grafana $CLUSTER` will do the same thing as `roachprod {start,stop}-grafana $CLUSTER$2`. That being said, roachtests should be able to set up whatever configuration they want by using the public methods in `pkg/roachprod/prometheus`.